feat: security section expansion (#229)
sparkbox/standard: security/security_principles/README.md

# Security Principles for Developers

> Security on the web relies on confidentiality, integrity, and availability. Considering each of these key pieces while we build helps us produce robust security control. The following security principles embody these ideals.

Reviewer: ```suggestion
Security on the web relies on confidentiality, integrity, and availability. We are able to produce robust security control by considering each of these key pieces while we build software. The following security principles embody these ideals.
```

**A secure application defines what is allowed and rejects everything else.**

For example, if we’re using the positive security model while working with a form input, we should specify only what characters will be allowed. That may mean only alphanumeric characters are allowed or only numbers are allowed, for example.

Another example is in user access control, where a user would be allowed access only to what is needed and denied access to everything else.

## Fail Securely

**A secure application handles errors and exceptions securely.**

Handling errors securely is a key aspect of secure coding. Exceptions that occur while processing a security control deserve special attention. These exceptions must not allow behavior that the control wouldn’t normally allow. Any time we implement a security mechanism, there are three possible outcomes:

Allow the operation
Disallow the operation
Exception

Generally, we design security mechanisms so that failure, or exception, is the same as disallowing the operation.

## Principle of Least Privilege

**A secure application gives the least amount of privilege required.**

This principle encompasses user rights and resource permissions.

> For example, when we add clients to Slack, we only give them access to one external client channel.
> 
> Another example is a server that only needs access to a network, read access to a database table, and the ability to write a log. So those are the only permissions that the server should be granted.

Reviewer: ```suggestion
Examples of least privilege include:

* only give clients access to one external client Slack channel
* only giving a server permission to a network, read access to a database table, and the ability to write a log
```

Keep in mind that it’s often easier to give additional permissions than to take away existing permissions.

## Separation of Duties

**A secure application separates duties, restricting the power of any individual.**

Separation of duties manages conflict of interest by restricting any one individual’s power. It is best implemented by using role-based access control and by carefully defining what each role should be able to do.

> We can apply this method to job roles, where a developer develops, a QA tester tests, and a devops engineer deploys. This kind of separation ensures plenty of eyes on the code before it goes live.

Reviewer: While the example applies on paper, the title-to-responsibility alignment is anathema to our principles. Everyone tests to create shared responsibility.

What if we used environments instead:

```suggestion
We can apply this method to environments, where a developers would not have administrative access to production environments. This kind of separation ensures production changes only occur through controlled means.
```

Author: TIL the word anathema. Great callout here.

Another example of this method is how we write code at Sparkbox. We write modular, componentized code to limit complexity, which can limit what an attacker may be able to communicate with in an attack.

## Security by Obscurity

> **A secure application is not reliant upon keeping details hidden.**

Reviewer: I _think_ we have this backwards. Security *by* obscurity means, don't tell folks what type of server or framework you are using. For example, it's best practice not to include the `X-Powered-By` header in HTTP repsonses.

Author: Really good point. I made some adjustments, including your example. I also tried to include some of [Bill's points](https://docs.google.com/document/d/1tQpVG4X4nvxfCGRz4hqcg1NN2nWxvnZVEwTpv_WJfoI/edit#bookmark=id.6zs408ieahah) about security by obscurity potentially being a weak control

Security through obscurity means “don’t try to hide things.” Attackers have all the time in the world to find things we hide from them. Security by obscurity is a weak security control, meaning that systems shouldn’t be reliant upon keeping details hidden.

For example, the security of a web application shouldn’t rely on knowledge of the source code being kept secret. That’s not to say keeping secrets is bad, but it’s not a reliable way to secure a web application on it’s own—be sure to use multiple levels of security on top of simply hiding information.

## Do Not Trust the Client

**A secure application validates information coming from the client.**

> This principle acknowledges that anything you send a client (a person or machine that’s consuming the application) can be changed and sent back. For this reason, validation should be performed in workflow or command patterns even if backend code performs validation as well. This validation process prevents unnecessary round trips to the backend and prevents the backend from revealing most tampered data.

Reviewer: The first statement is definitely accurate... what you send to the client can be tampered with. In fact, data sent to the server might not even come from your client code. (Postman is a good example of this).

For that reason, I think we can clarify this a little:

```suggestion
This principle acknowledges that anything you send a client (a person or machine that’s consuming the application) can be tampered with. Data sent from the client to the server should not be trusted. For this reason, the backend code should validate all data and permissions. 
```

I'm not sure I understand this.
> This validation process prevents unnecessary round trips to the backend and prevents the backend from revealing most tampered data.

Author: Yep this adjustment is way more clear 👍

Separation of duties manages conflict of interest by restricting any one individual’s power. It is best implemented by using role-based access control and by carefully defining what each role should be able to do.

We can apply this method to environments, where developers would not have administrative access to production environments. This kind of separation ensures production changes only occur through controlled means.

Another example of this method is how we write code at Sparkbox. We write modular, componentized code to limit complexity, which can limit what an attacker may be able to communicate with in an attack.

## Security by Obscurity

**A secure application is not reliant upon keeping details hidden.**

> Security by obscurity means don't reveal what type of server or framework you're using. For example, it's best practice not to include the `X-Powered-By` header in HTTP responses.

Reviewer: How about this:

> Security through obscurity means limiting what is revealed about about the design or implementation of a system. For example, it's best practice not to include the `X-Powered-By` header in HTTP responses.

Author: Yes! 💥

> Security by obscurity can be a weak control if you're trying to hide information that you've shared. Attackers have all the time in the world to find things we hide from them, so systems shouldn’t be reliant only on keeping details hidden.

Reviewer: I like this!
